### ISSUE TITLE
Automatically promote network that passes SPRT?

### ISSUE THREAD

**Issue by Azirine** [gcp/leela-zero#398]
4e4d09be has passed for a while and still isn't being used for training. Seems like automating this process would speed up training significantly.

**Comment by l1t1** [comment]
already used

**Comment by gcp** [comment]
This is being considered, but there's tricky cases to handle such as the first test passing with say 56%, and then immediately afterwards a new test passing with 65%...but against something that is now no longer the best network.

I suspect it will take some testing and debugging before that can be enabled.

**Comment by isty2e** [comment]
@gcp That's why I suggested to test multiple networks simultaneously and performing z-tests between SPRT-passing ones. That is maximally efficient, I think.

**Comment by gcp** [comment]
"Testing simultaneously" is a bit ambiguous. The networks don't come out of the trainer at the same time. Even if they did, they won't pass or fail SPRT at the same time.

**Comment by isty2e** [comment]
Yes, but sometimes they seem to be uploaded at the same time. For such cases it can be handy I think. Anyway, still we can save some time by performing z-tests between networks which have passed SPRT. Of course, the automation can be complicated... I don't know how they are supplied to the DB if there is any, but presumably we can label networks with the latest game in the training data, reserve some spaces before uploading, and determine the best performing one once everything is tested? I'm not sure about it actually.

**Comment by Dorus** [comment]
You should just test the network in order. So if you submit network A, B and C. First run A vs X till it pass or fails. If A passes, you schedule B vs A but if A fails you schedule B vs X instead.

**Comment by gcp** [comment]
The problem is that the matches aren't generated on the server, they're generated by the trainer, so the logic above isn't possible.

I guess in theory the trainer could just dumbly upload things and let the server work out the appropriate matches though.

**Comment by jjoshua2** [comment]
I think it has to be that way, that the server assigns tests. Look at fishtest for a good example of SPRT. To be included in master, a patch has to pass master. Tests have to be redone against latest master for another patch that looked promising. SPRT doesn't work to just pick the higher win % to know which was better. You would have to use a different statistical test for that.

**Comment by Dorus** [comment]
> The problem is that the matches aren't generated on the server, they're generated by the trainer, so the logic above isn't possible.

I was under the impression that the server send out match requests. So as soon as a client upload a game, the server can work out the win/loss rate, do appropriate SPRT, and start a new network if needed.

The only problem i see is that once all 250 clients are updated to the new network, you will have send out all 400 game requests before SPRT had the chance to pass or fail. Might be worth it to make only a certain % of clients to evaluation matches, and also select only quick clients, that way evaluation matches will run quickly and get the chance to cancel on time.

Might be an idea to let the client also submit the "ms/move" and "seconds/game" parameter on game requests.

**Comment by gcp** [comment]
In fishtest the server does not assign tests. The programmers set the tests, which is actually how it works in LZ too.

**Comment by gcp** [comment]
>I was under the impression that the server send out match requests.

The server sends out "games to play" to the clients, but the matches to run come from a database table, and entries are inserted into that table from my trainer or for example by @roy7 manually.

>Might be worth it to make only a certain % of clients to evaluation matches

Yes, we considered this, something like queuing no more than 100 games per network at a time.

**Comment by gcp** [comment]
>You should just test the network in order. First run A vs X till it pass or fails. 

I'm not sure this is optimal. That's why I said I think running X vs A, X vs B and X vs C simultaneously and picking the highest LLR may be better.

Basically it would play X vs A unless A's LLR drops below 0 (i.e. this will mean the score drops below 55%). Then it'd play X vs B until it passes or gets a worse score than A, and so on.

(This is - to some extent - how LZ's search for the best move works)

**Comment by Dorus** [comment]
Ah interesting consideration.

I guess it depend on what you wat to optimize. Fewer games wasted on evaluation or fast evaluation. I guess running them simultaneously is a good opportunity to send our more match requests while keeping the number that roll out on SPRT rejection low.

How about this proposal:

Start all matches simultaneous. As soon as a network is accepted (now called A) by SPRT you stop making requests on B vs X and C vs X and start on B vs A and C vs A.

For the B vs A games you can include the already played B vs X games, thus no need to rerun all 400, if you already did 50 you just consider those as if they where against A too.

You could even add a correction to those games based on how much A won against X, but my theory is that you wont even need that. If B is weaker it will not pass SPRT even with the already included score against X.  If B is stronger than A it would have passed SPRT earlier, but even if it didn't it will show its strength against A in the new matches.

**Comment by Dorus** [comment]
Oh, and giving it some more thoughts:

If you do want to compensate for A's higher skill, you could calculate elo instead of just SPRT acceptance. As soon as B is +35 elo on A based on the A vs X, B vs X and A vs B matches. Of course only do elo comparison after SPRT acceptance on B. That might be similar to the LLR solution you talk about.

**Comment by isty2e** [comment]
If we are to test networks simultaneously, I believe that the solution I suggested - z-test between SPRT passing ones - is nearly optimal.

**Comment by Dorus** [comment]
Are 3 way z-tests possible? I guess after a network is accepted we get more accurate measurements if we play against the new network from that point on.

I came up with elo because i know it ca do a 3 way test, but i'm sure there are better statistic options (my statistic education is slightly limited as you notice :) )

**Comment by jjoshua2** [comment]
We could ask on fishcooking about testing multiple patches at once. They
have some good statisticians. Elo is almost certainly not optimal since we
only need to know which is the best not the absolute scores of each.

Doesn't the latest AZ paper tell us that we don't need to worry about
regression that we can just use the latest network?
If so we don't need to be very rigorous in our testing and just favor the
most recent one. Something like pick the latest one that passed sprt.

 If we just pick the highest % it might not be fair since sprt could abort
early with a few lucky early wins. And have much higher than one that
started average or unlucky and did a more normal number of games. The
number of games to pass also does not tell how strong the patch was.



On Dec 16, 2017 11:44 AM, "Dorus" <notifications@github.com> wrote:

Are 3 way z-tests possible? I guess after a network is accepted we get more
accurate measurements if we play against the new network from that point on.

I came up with elo because i know it ca do a 3 way test, but i'm sure there
are better statistic options (my statistic education is slightly limited as
you notice :) )

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/gcp/leela-zero/issues/398#issuecomment-352194645>, or mute
the thread

**Comment by isty2e** [comment]
​@Dorus The winrates of A-X and B-X can be compared. Considering that both should pass SPRT, we can save some games by directly comparing their winrates only.

**Comment by jjoshua2** [comment]
I think to compare the winrates you should have each play at least 200
games and have them all have played about the same number for it to be a
fair test that is mostly reliable.

Going with the previous idea that they all passed and it doesn't matter as
much which one you pick as the much as using a better one to learn early we
could maybe use the fishtest idea of two phase testing. Playouts at 400 or
800 would go 2 or 4x as fast and I think one of the papers had 800 Playouts
used.

We could only use the fewer Playouts when decided which of the ones that
passed to use or as a fast screen to do a full sprt test (would probably
have to much faster to save time like 5x).

**Comment by isty2e** [comment]
I have no idea why you consider the z-test less reliable... Basically you don't need to wait until 200 games are gathered. If something is statistically significant (within a certain p-value), that's it.

**Comment by jjoshua2** [comment]
I don't know how to do a n way ztest. Do you have code or a link to the
algorithm? I wasn't referring to that way being bed but to just using the
sprt results and picking the highest win rate.

**Comment by isty2e** [comment]
For example, [this](https://github.com/sethtroisi/leela-zero/blob/analysis/scripts/thread_compare.py) is a code to perform a two proportion z-test, implemented by @sethtroisi.

**Comment by jjoshua2** [comment]
Nice. It looks like at least 30 games is needed for a z test. But it looks
it only compared two? So how do you do 3 or 4 at a time? Are you taking the
one with the highest confidence of being better than master?

**Comment by isty2e** [comment]
@jjoshua2 We can perform a series of pairwise comparisons to rank networks then. No difficulty here.

**Comment by evanroberts85** [comment]
If you pick a 60% network over a 59% network, with them both up against the same current best, then it brings into question demaning 95 percent confidence in the first place, as the 60% networks is not 95% likely to be better than the 59% one. Simply changing the order they are plahed to stop the 59% one becoming best first does not really solve this question, it just avoids it!

I think really you should accept any network that is scoring 50% (or maybe 50.5%) of above, with a test of 95% confidence test that is is not significantly worse (say 35 elo)

**Comment by odeint** [comment]
When results are similar against the previous champion, I think it makes sense to use the number of training steps as a tie-breaker. The network with fewer training steps is more likely to not be overfitted and have learnt something general. We already do this implicitly by usually uploading the networks in ascending training step order, so that the one with fewer steps has the chance to become the new "king of the hill" first.

**Comment by roy7** [comment]
Yes I think that is why @gcp uploads in that order and asked me to do testing sequentially, at least for now. Uploading multiple networks at once is more in artifact that he is doing it by hand, then an attentional decision. If we get networks to test automated, then they'll be coming in on a periodic basis. So the odds are a new network will get fully tested before another network is uploaded.

**Comment by Dorus** [comment]
@evanroberts85 But at 400 games, you need 55% win to be 95% confident the new network is scoring at least 50%. When picking between 2 new networks, all we really care about is that it is not worse than the previous one. Even if by random luck we pick the one that scored 1% better but is really 1% worse, that wont really affect our eventual outcome. So we shouldn't care too much.

**Comment by evanroberts85** [comment]
My argument was by demanding a 95% confidence, each time a network does not
pass, It means you are preferring a network that has a up to 95% chance of
being worse than the new network. This is a bias of the older network over
the new and I can see no logical reason for having this bias, if anything
it should be the other way round.

**Comment by gcp** [comment]
Doing a series of z tests on multiple combinations ruins their 95% guarantee unless you discount for the multiple evaluations. It's almost guaranteed to be highly suboptimal because of that.

**Comment by isty2e** [comment]
That is true to an extent, but usually 3 or 4 networks are given at a time, and only one or two pass the SPRT, so it wouldn't be a problem in practice. Even if so, we can lower p-value accordingly in such cases.

**Comment by DanielThomas55** [comment]
https://www.remi-coulom.fr/WHR/

**Comment by gcp** [comment]
There's no history on any of the players evaluated, so that's rather pointless.

**Comment by gcp** [comment]
This was implemented a while ago.
